netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -801,94 +852,239 @@ _net_ebpf_extension_connection_context_initialize(
+        if (!delete_all && entry->timestamp > expiry_time) {
+            break;
+        }
+
+#pragma warning(suppress : 6001) /* entry and list entry are non-null */
+        list_entry = list_entry->Blink;
+        RemoveEntryList(&entry->list_entry);
+
+        if (delete_all) {
+            // Free the memory
+            ExFreePool(entry);
+        } else {
+            // Return the entry to the free list
+            InsertHeadList(&_net_ebpf_ext_low_memory_free_connect_context_list, &entry->list_entry);
+        }
+    }
+
+    // Free pre-allocated entries if we are shutting down.
+    if (delete_all) {
+        while (!IsListEmpty(&_net_ebpf_ext_low_memory_free_connect_context_list)) {
+            PLIST_ENTRY entry = RemoveHeadList(&_net_ebpf_ext_low_memory_free_connect_context_list);
+            net_ebpf_extension_connection_context_t* context =
+                CONTAINING_RECORD(entry, net_ebpf_extension_connection_context_t, list_entry);
+            ExFreePool(context);
+        }
+    }
+
     NET_EBPF_EXT_LOG_MESSAGE_UINT64(
         NET_EBPF_EXT_TRACELOG_LEVEL_INFO,
         NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
-        "_net_ebpf_ext_purge_lru_contexts_under_lock",
+        "_net_ebpf_ext_purge_connect_contexts_under_lock",
         _net_ebpf_ext_connect_context_count);
 }
 
 static void
-_net_ebpf_ext_purge_lru_contexts(BOOLEAN delete_all)
+_net_ebpf_ext_purge_connect_contexts(BOOLEAN delete_all)
 {
     KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
-    _net_ebpf_ext_purge_lru_contexts_under_lock(delete_all);
+
+    _net_ebpf_ext_purge_connect_contexts_under_lock(delete_all);
+    if (delete_all) {
+        ebpf_hash_table_destroy(_net_ebpf_ext_blocked_connect_context_hash_table);
+        _net_ebpf_ext_blocked_connect_context_hash_table = NULL;
+    }
+
     ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
 }
 
-static void
-_net_ebpf_ext_insert_connection_context_to_list(_Inout_ net_ebpf_extension_connection_context_t* connection_context)
+_Requires_exclusive_lock_held_(_net_ebpf_ext_sock_addr_lock) static ebpf_result_t
+    _net_ebpf_ext_insert_connection_context_to_low_memory_list(
+        _In_ const bpf_sock_addr_t* sock_addr_ctx, UINT64 transport_endpoint_handle)
 {
-    KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+    ebpf_result_t result = EBPF_SUCCESS;
+    PLIST_ENTRY entry = NULL;
+    net_ebpf_extension_connection_context_t* blocked_connection_context = NULL;
+    if (IsListEmpty(&_net_ebpf_ext_low_memory_free_connect_context_list)) {
+        NET_EBPF_EXT_BAIL_ON_ERROR_RESULT(EBPF_NO_MEMORY);
+    }
 
-    // Insert the most recent entry at the head.
-    InsertHeadList(&_net_ebpf_ext_connect_context_list, &connection_context->list_entry);
-    _net_ebpf_ext_connect_context_count++;
+    // Retrieve an entry from the pre-allocated list
```

Comment:
```suggestion
    // Retrieve an entry from the pre-allocated list.
```
(most comments end in punctuation, but many of the new ones added in this PR do not. Use consistent style.